packages/serverless-orchestration/src/ServerlessHub.js

```diff
@@ -64,29 +66,98 @@ hub.post("/", async (req, res) => {
     }
     // Get the config file from the GCP bucket if running in production mode. Else, pull the config from env.
     const configObject = await _fetchConfig(req.body.bucket, req.body.configFile);
-
-    // Fetch the last block number this given config file queried the blockchain at if running in production. Else, pull from env.
-    const lastQueriedBlockNumber = await _getLastQueriedBlockNumber(req.body.configFile);
-    if (!configObject || !lastQueriedBlockNumber)
+    if (!configObject)
       throw new Error(
-        `Serverless hub requires a config object and a last updated block number! configObject:${JSON.stringify(
-          configObject
-        )} lastQueriedBlockNumber:${lastQueriedBlockNumber}`
+        `Serverless hub missing a config object! GCPBucket:${req.body.bucket} configFile:${req.body.configFile}`
       );
+    logger.debug({
+      at: "ServerlessHub",
+      message: "Executing Serverless query from config file",
+      spokeUrl,
+      botsExecuted: Object.keys(configObject),
+    });
+
+    // As a first pass, loop over all config objects in the config file and fetch the last queried block number and
+    // head block for each unique chain ID. The reason why we precompute these block numbers for each chain ID is so
+    // that each bot connected to the same chain will use the same block number parameters, which is a convenient
+    // assumption if there are many bots running on the same chain.
+    let blockNumbersForChain = {
+      // (chainId: int): {
+      //     lastQueriedBlockNumber: <int>
+      //     latestBlockNumber: <int>
+      // }
+    };
+    for (const botName in configObject) {
+      // Check if bot is running on a non-default chain, and then fetch last block number seen on this or the default
+      // chain.
+      const chainId = configObject[botName]?.gcpDataStoreChainId || DEFAULT_GCP_DATASTORE_CHAINID;
+
+      // If we've seen this chain ID already we can skip it:
+      if (blockNumbersForChain[chainId]) continue;
+
+      // Fetch last seen block for this chain:
+      let lastQueriedBlockNumber = await _getLastQueriedBlockNumber(req.body.configFile, chainId);
+
+      // Next, get the head block for the chosen chain, which we'll use to override the last queried block number
+      // stored in GCP at the end of this hub execution.
+      let latestBlockNumber;
+
+      // If bot wants to use a non-default chain, then it must also specify a `CUSTOM_NODE_URL`, which we'll use to
+      // fetch the head block for the chain.
+      const spokeCustomNodeUrl = configObject[botName]?.environmentVariables?.CUSTOM_NODE_URL;
```

Comment:
This bit of code makes a subtle assumption that the `CUSTOM_NODE_URL` and the `gcpDataStoreChainId` are matched. I don't love making this assumption, but I'm not sure how else to dynamically grab a node URL for the chosen chainId without using the bot's custom set node URL.

Reply:
Hmmm, why do we even need the `gcpDataStoreChainId`? Could we:
1. Check for a CUSTOM_NODE_URL -- if one doesn't exist use the hub's CUSTOM_NODE_URL.
2. Get the chain id for that node. (web3.eth.getChainId)
3. Get the block number for that node.
4. Store them.

If you want to avoid excess queries and overwriting, you could construct a local cache of node url to chain id and block number, but that's just an optimization.

Thoughts?